### PR TITLE
Converting Glib's mutex (obsolete) to std::mutex

### PR DESCRIPTION
**Pull request by Hombre57**

### CODE REVIEW

**Comment by Hombre57** [comment]
@Floessie I've made the conversion and didn't had performance loss or deadlock so far. I think that the conversion was pretty straightforward, but I prefer the review of an experienced dev :wink: . The review is about the `Glib` vs `std` equivalence, not the utility class as a whole, which anyone can deeply review on a later date...

**Comment by heckflosse** [comment]
Without looking at the changes, I would see that for RT 5.8
Though I will @Floessie have the last word on this...

File: rtgui/threadutils.cc
Comment (Floessie) [inline rtgui/threadutils.cc]: Hm, we want multiple readers, so we should notify them all, don't we?

I asked my workmate to review this piece of code, too, and he noticed a possible imbalance in `MyReaderLock::acquire()`:

```c++
void MyReaderLock::acquire ()
{
    if (locked) {
        return;
    }

    std::lock_guard<std::mutex> lock (mutex.mutex);
// Precondition: A writer is writing, so writerCount != 0.
    if (mutex.writerCount == 0) {
        // There's no writer operating, we can increment the writer count which will lock writers.
        ++mutex.writerCount;
    } else if (mutex.readerCount == 0) {
// There is no active reader yet, so readerCount == 0.
        // The writer count is non null, but a reader can be the owner of the writer lock,
        // which will be the case if the reader count is not zero too.
// Now let two readers spin here.
        while (mutex.writerCount != 0) {
            mutex.cond.wait(mutex.mutex);
        }
// The writer leaves, both readers proceed.
        // Then, we can increment the writer count.
// The writerCount is incremented by 2. This leads to an imbalance in release().
        ++mutex.writerCount;
    }

    // Finally, we can increment the reader count as well.
    ++mutex.readerCount;

    locked = true;
}

void MyReaderLock::release ()
{
    if (!locked) {
        return;
    }

    std::lock_guard<std::mutex> lock (mutex.mutex);

    // decrement the writer number first...
    --mutex.readerCount;

    if (mutex.readerCount == 0) {
        // ...if no more reader, we decrement the writer count as well...
// Last reader leaves, but writerCount is only decremented by 1.
        --mutex.writerCount;

        // ...and signal the next waiting reader/writer that it's free
        mutex.cond.notify_one();  // notify_all ?
    }

    locked = false;
}
```

Is this plausible?

File: rtgui/threadutils.h
Comment (Floessie) [inline rtgui/threadutils.h]: [cppreference](https://en.cppreference.com/w/cpp/thread/condition_variable_any) states:
> If the lock is std::unique_lock, std::condition_variable may provide better performance.

Maybe something like this?

```patch
diff --git a/rtgui/threadutils.cc b/rtgui/threadutils.cc
index 4f8190c36..b7e18e955 100644
--- a/rtgui/threadutils.cc
+++ b/rtgui/threadutils.cc
@@ -68,7 +68,7 @@ void MyReaderLock::acquire ()
         return;
     }
 
-    std::lock_guard<std::mutex> lock (mutex.mutex);
+    std::unique_lock<std::mutex> lock (mutex.mutex);
 
     if (mutex.writerCount == 0) {
         // There's no writer operating, we can increment the writer count which will lock writers.
@@ -77,7 +77,7 @@ void MyReaderLock::acquire ()
         // The writer count is non null, but a reader can be the owner of the writer lock,
         // which will be the case if the reader count is not zero too.
         while (mutex.writerCount != 0) {
-            mutex.cond.wait(mutex.mutex);
+            mutex.cond.wait(lock);
         }
 
         // Then, we can increment the writer count.
@@ -96,7 +96,7 @@ void MyReaderLock::release ()
         return;
     }
 
-    std::lock_guard<std::mutex> lock (mutex.mutex);
+    const std::unique_lock<std::mutex> lock (mutex.mutex);
 
     // decrement the writer number first...
     --mutex.readerCount;
@@ -118,11 +118,11 @@ void MyWriterLock::acquire ()
         return;
     }
 
-    std::lock_guard<std::mutex> lock (mutex.mutex);
+    std::unique_lock<std::mutex> lock (mutex.mutex);
 
     // The writer count is not zero, so we have to wait for it to be zero again...
     while (mutex.writerCount != 0) {
-        mutex.cond.wait (mutex.mutex);
+        mutex.cond.wait (lock);
     }
 
     // ...then we can increment the writer count.
@@ -137,7 +137,7 @@ void MyWriterLock::release ()
         return;
     }
 
-    std::lock_guard<std::mutex> lock (mutex.mutex);
+    const std::unique_lock<std::mutex> lock (mutex.mutex);
 
     // Decrement the writer number first...
     if (--mutex.writerCount == 0) {
diff --git a/rtgui/threadutils.h b/rtgui/threadutils.h
index f54cbb845..9a7e810d8 100644
--- a/rtgui/threadutils.h
+++ b/rtgui/threadutils.h
@@ -91,7 +91,7 @@ public:
 
 private:
     std::mutex mutex;
-    std::condition_variable_any cond;
+    std::condition_variable cond;
 
     std::size_t writerCount = 0;
     std::size_t readerCount = 0;
```

File: rtgui/threadutils.h
Comment (Floessie) [inline rtgui/threadutils.h:208]: :+1: for removing unused code.

File: rtgui/thumbimageupdater.cc
Comment (Floessie) [inline rtgui/thumbimageupdater.cc]: Don't know. Is a thundering herd allowed here?

File: rtgui/thumbimageupdater.cc
Comment (Floessie) [inline rtgui/thumbimageupdater.cc]: Again, a `std::condition_variable` could be faster.

**Review by Floessie — CHANGES_REQUESTED** [review]

**Comment by Hombre57** [comment]
Thanks for the review @Floessie , I'll get back on this this w.e.

**Comment by Beep6581** [comment]
Do we merge?

File: rtgui/threadutils.cc
Comment (hfiguiere) [inline rtgui/threadutils.cc]: this shouldn't be added if you remove the `Glib::Threads` (here it builds without)

**Comment by hfiguiere** [comment]
How can I help getting this merged? 

I was about to prepare something similar. This would greatly help build the rtengine with a more modern gilbmm.

File: rtgui/threadutils.cc
Comment (hfiguiere) [reply rtgui/threadutils.cc]: `broadcast`  notifies all waiting threads, so `notify_all` is likely the proper answer.

File: rtgui/thumbimageupdater.cc
Comment (hfiguiere) [reply rtgui/thumbimageupdater.cc]: `notify_all` is the proper equivalent here.

**Comment by Hombre57** [comment]
Damn ! How can some dude start something useful and almost ready and leave the sh!t undone ? I apologize for this stalled branch, I'll resurrect it ASAP.

**Comment by ghost** [comment]
Hello Hombre,

do you need any help? If so, just give me a short hint. After the switch to std has been merged, I will got on with a look at gtk4.

Thank you in advance.

**Comment by Hombre57** [comment]
Hi Christian,

Hopefully I'll be able to work on this tomorrow. I'll ask for your (and others) help here if needed. I'll need reviewer anyway :wink:

File: rtgui/threadutils.cc
Comment (Hombre57) [reply rtgui/threadutils.cc]: @Floessie 

About 
```
// Now let two readers spin here.
```

Isn't `std::lock_guard<std::mutex> lock (mutex.mutex);` protecting from this possibility ?

**Comment by Hombre57** [comment]
I've made the changes locally, I'll have one next try for pushing the changes to `std-mutex` branch, unless I don't have commit rights anymore. For now it send's me back `not authorized`, but I don't know if it's related to some SSH public key trouble or repositories access rights. Hi @Beep6581 :slightly_smiling_face: , any idea ?

**Comment by Hombre57** [comment]
Yay ! I managed to find the correct connection string in Eclipse, now waiting for your review.

**Comment by ghost** [comment]
Hello everyone,

thank you so much for pushing this Hombre57. As far as I can judge - based on my knowledge which is limited to thread handling in gtk/gdk - it looks ok. Most of the changes have also been done and tested in my repo (the PL has been closed cause I have not seen, that one exists already here).

As far as I can see here, the review of Floessie is missing to get the commits into dev right?

Greetings

**Comment by Hombre57** [comment]
Hi Christian,

I think I've done things the wrong way. I've added @Floessie's requested changes in my commits w/o validating it here. So I'll dismiss it but, in fact, the changes have been taken into account.

**Comment by Hombre57** [comment]
Sorry but I'll commit this PR, it integrates the requested changes. I've used this branch for some times now w/o problem. In any case, a fix should be straightforward.

**Pull request merged**
